java/ql/src/experimental/Security/CWE/CWE-749/UnsafeAndroidAccess.ql

```diff
@@ -0,0 +1,139 @@
+/**
+ * @name Unsafe resource loading in Android webview
+ * @description JavaScript rendered inside WebViews can access any protected application file and web resource from any origin
+ * @kind path-problem
+ * @tags security
+ *       external/cwe/cwe-749
+ *       external/cwe/cwe-079
+ */
+
+import java
+import semmle.code.java.frameworks.android.Intent
+import semmle.code.java.frameworks.android.WebView
+import semmle.code.java.dataflow.FlowSources
+
+/**
+ * Allow universal access methods in the WebSettings class
+ */
+class AllowUniversalAccessMethod extends Method {
+  AllowUniversalAccessMethod() {
+    this.getDeclaringType() instanceof TypeWebSettings and
+    (
+      this.hasName("setAllowUniversalAccessFromFileURLs") or
+      this.hasName("setAllowFileAccessFromFileURLs")
+    )
+  }
+}
+
+/**
+ * `setJavaScriptEnabled` method for the webview
+ */
+class AllowJavaScriptMethod extends Method {
+  AllowJavaScriptMethod() {
+    this.getDeclaringType() instanceof TypeWebSettings and
+    this.hasName("setJavaScriptEnabled")
+  }
+}
+
+/**
+ * Check whether JavaScript is enabled in the webview with universal resource access
+ */
+predicate isJSEnabled(MethodAccess ma) {
+  exists(VarAccess va, MethodAccess jsa |
+    ma.getQualifier() = va and
+    jsa.getQualifier() = va.getVariable().getAnAccess() and
+    jsa.getMethod() instanceof AllowJavaScriptMethod and
+    jsa.getArgument(0).(BooleanLiteral).getBooleanValue() = true
+  )
+}
+
+/**
+ * Load URL method call on the `android.webkit.WebView` object
+ */
+class LoadResourceMethodAccess extends MethodAccess {
+  LoadResourceMethodAccess() {
+    this.getMethod().getDeclaringType() instanceof TypeWebView and
+    (
+      this.getMethod().hasName("loadUrl") or
+      this.getMethod().hasName("postUrl")
+    )
+  }
+}
+
+/**
+ * Method access to external inputs of `android.content.Intent` object
+ */
+class IntentGetExtraMethodAccess extends MethodAccess {
+  IntentGetExtraMethodAccess() {
+    this.getMethod().getName().regexpMatch("get\\w+Extra") and
+    this.getMethod().getDeclaringType() instanceof TypeIntent
+    or
+    this.getMethod().getName().regexpMatch("get\\w+") and
+    this.getQualifier().(MethodAccess).getMethod().hasName("getExtras") and
+    this.getQualifier().(MethodAccess).getMethod().getDeclaringType() instanceof TypeIntent
+  }
+}
+
+/**
+ * Source of loading urls
+ */
+class UntrustedResourceSource extends RemoteFlowSource {
+  UntrustedResourceSource() {
+    exists(MethodAccess ma |
+      ma instanceof IntentGetExtraMethodAccess and
+      this.asExpr().(VarAccess).getVariable().getAnAssignedValue() = ma
+    )
+  }
+
+  override string getSourceType() {
+    result = "user input vulnerable to XSS and sensitive resource disclosure attacks" and
+    exists(MethodAccess ma |
+      ma.getMethod() instanceof AllowUniversalAccessMethod and //High precision match of unsafe resource loading
+      ma.getArgument(0).(BooleanLiteral).getBooleanValue() = true
+    )
```

Comment:
Factor out repeated code

Reply:
@smowton The reason I have this code snippet is to output different messages for two scenarios - with the <code>CrossOriginAccessMethod</code> set to true or not. For the former, there are more severe security vulnerabilities when it's set to true. And it's still possible to have vulnerabilities like XSS otherwise. That's what I mean "high precision match".

Is there a better way to model this? Thanks.

Reply:
Thanks a lot for your help, I've added the new implementation.